javascript/ql/src/semmle/javascript/security/dataflow/Xss.qll

```diff
@@ -22,6 +22,23 @@ module Shared {
 
   /** A sanitizer for XSS vulnerabilities. */
   abstract class Sanitizer extends DataFlow::Node { }
+
+  /**
+   * A regexp replacement involving an HTML meta-character, viewed as a sanitizer for
+   * XSS vulnerabilities.
+   *
+   * The XSS queries do not attempt to reason about correctness or completeness of sanitizers,
+   * so any such replacement stops taint propagation.
+   */
+  class MetacharEscapeSanitizer extends Sanitizer, DataFlow::MethodCallNode {
+    MetacharEscapeSanitizer() {
+      getMethodName() = "replace" and
+      exists(RegExpConstant c |
+        c.getLiteral() = getArgument(0).asExpr() and
+        c.getValue().regexpMatch("['\"&<>]")
```

Comment:
Could we avoid matching on the single and double quotes? I fear they will make too many false XSS sanitizers since I think it is common to strip or replace one quote type with the other. This change motivates a slight renaming of the class to reflect that it does not identify all meta-characters.

Besides, the single quote is not a meta character, unless the browser supports it as an attribute value delimiter for compatibility reasons. I think.

Reply:
Sure, I'll take out the quotes.

> the single quote is not a meta character

https://www.w3.org/TR/2012/WD-html-markup-20120320/syntax.html#syntax-attr-single-quoted

Reply:
Ugh, no, wait, I can't do that, otherwise it'll start flagging uses of `verifyURL` from #205 again...

Reply:
Ok then.